xptifw/src/xpti_trace_framework.cpp

```diff
@@ -0,0 +1,1204 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+#include "xpti_trace_framework.hpp"
+#include "xpti_int64_hash_table.hpp"
+#include "xpti_string_table.hpp"
+
+#include <cassert>
+#include <map>
+#include <mutex>
+#include <sstream>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#ifdef XPTI_STATISTICS
+#include <stdio.h>
+#endif
+
+#ifdef XPTI_USE_TBB
+#include <tbb/concurrent_hash_map.h>
+#include <tbb/concurrent_unordered_map.h>
+#include <tbb/concurrent_vector.h>
+#include <tbb/spin_mutex.h>
+#endif
+
+#define XPTI_USER_DEFINED_TRACE_TYPE16(value)                                  \
+  ((uint16_t)xpti::trace_point_type_t::user_defined | (uint16_t)value)
+#define XPTI_USER_DEFINED_EVENT_TYPE16(value)                                  \
+  ((uint16_t)xpti::trace_event_type_t::user_defined | (uint16_t)value)
+#define XPTI_EXTRACT_MSB16(val) (val >> 16)
+#define XPTI_EXTRACT_LSB16(val) (val & 0x0000ffff)
+
+#define XPTI_VENDOR_DEFINED_TRACE_TYPE16(vendor_id, trace_type)                \
+  ((uint16_t)vendor_id << 8 | XPTI_USER_DEFINED_TRACE_TYPE16(trace_type))
+#define XPTI_VENDOR_DEFINED_EVENT_TYPE16(vendor_id, event_type)                \
+  ((uint16_t)vendor_id << 8 | XPTI_USER_DEFINED_EVENT_TYPE16(event_type))
+
+namespace xpti {
+constexpr const char *env_subscribers = "XPTI_SUBSCRIBERS";
+xpti::utils::PlatformHelper g_helper;
+// This class is a helper class to load all the listed subscribers provided by
+// the user in XPTI_SUBSCRIBERS environment variable.
+class Subscribers {
+public:
+  // Data structure to hold the plugin related information, including the
+  // initialization and finalization functions
+  struct plugin_data_t {
+    /// The handle of the loaded shared object
+    xpti_plugin_handle_t handle = nullptr;
+    /// The initialization entry point
+    xpti::plugin_init_t init = nullptr;
+    /// The finalization entry point
+    xpti::plugin_fini_t fini = nullptr;
+    /// The name of the shared object (in UTF8?))
+    std::string name;
+    /// indicates whether the data structure is valid
+    bool valid = false;
+  };
+  // Data structures defined to hold the plugin data that can be looked up by
+  // plugin name or the handle
+  //
+  using plugin_handle_lut_t = std::map<xpti_plugin_handle_t, plugin_data_t>;
+  using plugin_name_lut_t = std::map<std::string, plugin_data_t>;
+
+  // We unload all loaded shared objects in the destructor; Must not be invoked
+  // in the DLLMain() function and possibly the __fini() function in Linux
+  ~Subscribers() { unloadAllPlugins(); }
+  // Method to query the plugin data information using the handle. If there's no
+  // information present for the handle provided, the method returns a structure
+  // with the valid attribute set to 'false'
+  plugin_data_t queryPlugin(xpti_plugin_handle_t Handle) {
+    plugin_data_t PData;
+#ifdef XPTI_USE_TBB
+    tbb::spin_mutex::scoped_lock MyLock(MMutex);
```

Comment:
Perhaps you could suggest the TBB team to provide a standard mutex API?

Reply:
@keryell  For the critical section sizes here, the spin locks give better performance with increasing number of threads - as in the events/sec remain relatively constant whereas the std::mutex tends to drop off when threads are > 4ish. I am not sure if there are use cases that enqueue kernels from 10s of threads, but it is nice to have an option to handle that nicely, if required. This is the reason why std containers are the default and if required, users can enable TBB.

Reply:
Ah sure. I agree with that. I was really talking about the API, not the mutex implementation details or semantics.
For example it would be nice to use it with:
```c++
std::unique_lock<tbb::spin_mutex> MyLock(MMutex);
```

Reply:
Ah, I see it now - I'll definitely suggest this to the team.